### PR TITLE
Generate public API listings and add up-to-date check for them

### PR DESCRIPTION
**Pull request by pakrym**
For all Track 2 libraries so API changes can be reviewed easier.

This is an exploratory PR, tell me what you think.

### CODE REVIEW

**Review by jsquire — APPROVED** [review]
With respect to Event Hubs, there's not much value in generating the API stubs at present since we're in the middle of implementing a fairly wholesale revamp of the entire API surface.

**Comment by pakrym** [comment]
> With respect to Event Hubs, there's not much value in generating the API stubs at present since we're in the middle of implementing a fairly wholesale revamp of the entire API surface.

Sure, the goal is to avoid breaking libraries that went GA it just works for everyone and can be disabled on per-project basis.

File: eng/ApiListing.targets [hunk 1]
```diff
@@ -0,0 +1,53 @@
+<Project>
+
+  <Target Name="GenerateApiListing">
```
Comment (weshaggard) [inline eng/ApiListing.targets]: Why generate and check the api ref? Is it just a caching mechanism? We could also pull down the last package and do a diff and api check. I was thinking that is likely what we would do when we enable the api breaking change checking.

File: eng/ApiListing.targets [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    </ItemGroup>
+
+    <MSBuild Projects="$(MSBuildProjectFullPath)"
+             Targets="_GenerateApiListingInner"
+             Properties="TargetFramework=%(_DeduplicatedTargetFramework.Identity)">
+    </MSBuild>
+
+  </Target>
+
+  <Target Name="_GenerateApiListingInner" DependsOnTargets="Build" Condition="'$(GenerateAPIListing)' == 'true'">
```
Comment (weshaggard) [inline eng/ApiListing.targets]: I'm generally not a fan of mutating the source code in the source tree during the build automatically. If we did this I think we should consider generating it to the obj folder and then doing a diff or something similar.

File: eng/ApiListing.targets (hunk 1 above)
Comment (pakrym) [reply eng/ApiListing.targets]: This is something I knew how to do :) 

Yes, I check them in as package mechanism as pulling the previous version of package requires defining what "previous" is and keeping that up-to-date etc.

I'm happy to switch to that when we enable breaking change checking but wanted something quick because new changes to GA libraries are coming in

File: eng/ApiListing.targets (hunk 2 above)
Comment (pakrym) [reply eng/ApiListing.targets]: It's not during the build, developer has to run eng\Generate-API to update this code. And there is a CI check to verify it's up-to-date.

File: eng/CodeChecks.ps1 [hunk 3]
```diff
@@ -71,7 +76,7 @@ try {
     if ($LastExitCode -ne 0) {
         $status = git status -s | Out-String
         $status = $status -replace "`n","`n    "
-        LogError "Generated code is not up to date. You may need to run eng\Update-Snippets.ps1 or sdk\storage\generate.ps1"
+        LogError "Generated code is not up to date. You may need to run eng\Update-Snippets.ps1 or sdk\storage\generate.ps1 or Generate-API.ps1"
```
Comment (heaths) [inline eng/CodeChecks.ps1]: Nit: use commas between articles, i.e. "eng\Update-Snippets.ps1, sdk\storage\generate.ps1, or Generate-API.ps1"

But what's the path for Generate-API? Also, that's a non-standard verb. Probably won't make a difference depending on strictness of script hosts, but some will throw when importing. Might use Extract-API or something instead.

File: eng/Directory.Build.Data.targets [hunk 4]
```diff
@@ -43,6 +45,11 @@
     <None Include="$(RepoEngPath)/images/pkgicon.png" Pack="true" PackagePath=""/>
   </ItemGroup>
 
+  <!-- Don't compile API listing files -->
+  <ItemGroup Condition="'$(GenerateAPIListing)' == 'true'">
+    <Compile Remove="api/*.cs" />
```
Comment (heaths) [inline eng/Directory.Build.Data.targets]: Does this actually match anything? Below, "api" is sibling to "src" and "test", so why would it be a potential Compile item anyway?

File: eng/Generate-API.ps1 [hunk 5]
```diff
@@ -0,0 +1,3 @@
+$servicesProj = Resolve-Path "$PSScriptRoot/service.proj"
+
+dotnet msbuild /p:Configuration=Release /restore /t:GenerateAPIListing $servicesProj
```
Comment (heaths) [inline eng/Generate-API.ps1]: Shouldn't this be "dotnet build ..."?

File: eng/Packages.Data.props [hunk 6]
```diff
@@ -111,6 +111,7 @@
     <PackageReference Update="Microsoft.Extensions.DependencyInjection" Version="2.1.0" />
     <PackageReference Update="Microsoft.Extensions.Logging.Abstractions" Version="2.1.0" />
     <PackageReference Update="Microsoft.Extensions.Options" Version="2.1.0" />
+    <PackageReference Update="Microsoft.DotNet.GenAPI" Version="5.0.0-beta.19552.1" />
```
Comment (heaths) [inline eng/Packages.Data.props]: Aside: huh, look at that: "beta". ;)

**Review by heaths — APPROVED** [review]

File: eng/ApiListing.targets (hunk 2 above)
Comment (pakrym) [reply eng/ApiListing.targets]: I don't want to slow down build for the rare times public API changes are made. CI checks would remind them :)

File: eng/CodeChecks.ps1 (hunk 3 above)
Comment (pakrym) [reply eng/CodeChecks.ps1]: I'm fine with Extract-API

File: eng/Generate-API.ps1 (hunk 5 above)
Comment (pakrym) [reply eng/Generate-API.ps1]: No, `dotnet msbuild` is the right thing here, I'm not trying to build, just invoke the custom target.

File: eng/Directory.Build.Data.targets (hunk 4 above)
Comment (pakrym) [reply eng/Directory.Build.Data.targets]: Ha, good point, I thought I've put the generated file inside the src directory but now I kind of like having it top level.

**Review by tg-msft — APPROVED** [review]
How do we disable this?  I'd like to pull DataLake as well since it'll be going through a lot of changes.

**Comment by pakrym** [comment]
I'll make a change it for EventHubs and DataLake

`<GenerateAPIListing>false</GenerateAPIListing>` in the project file should do it.

File: eng/Generate-API.ps1 (hunk 5 above)
Comment (heaths) [reply eng/Generate-API.ps1]: Sure, but you can pass targets (/t) to build as well. I suppose as long as dependencies are restored in either case, it really doesn't matter.

**Comment by heaths** [comment]
We may need that for Certificates, too. But I should be done with API changes. I'll take care of adding it if I need to.

**Comment by pakrym** [comment]
cc @AlexGhiondea

File: eng/ApiListing.targets [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      <_RefProjectFileTFM>$(TargetFramework)</_RefProjectFileTFM>
+      <_RefProjectFileTFM Condition="'$(TargetFramework)' == '$(DefaultNetCoreTargetFramework)'">%24(DefaultNetCoreTargetFramework)</_RefProjectFileTFM>
+
+      <_RefSourceOutputPath>$([System.IO.Directory]::GetParent('$(MSBuildProjectDirectory)'))/api/</_RefSourceOutputPath>
+      <_RefSourceFileName>$(AssemblyName).$(_RefSourceFileTFM).cs</_RefSourceFileName>
+      <_RefSourceFileOutputPath>$(_RefSourceOutputPath)$(_RefSourceFileName)</_RefSourceFileOutputPath>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <_ReferenceDirectoriesWithDuplicates Include="@(ReferencePath->'%(RootDir)%(Directory)'->TrimEnd('\'))" />
```
[the remark this answers is not in the document]
Comment (pakrym) [reply eng/ApiListing.targets]: Yes, this code trims the slashes. What would be atypical?

File: eng/ApiListing.targets (hunk 2 above)
Comment (weshaggard) [reply eng/ApiListing.targets]: Where is `eng\Generate-API.ps1`?

Comment (pakrym) [reply eng/ApiListing.targets]: Renamed to Export-API

Comment (pakrym) [reply eng/ApiListing.targets]: https://github.com/Azure/azure-sdk-for-net/pull/8695/files#diff-3d8d4efb9238e6d9bd7e0ebe9d325476

Comment (weshaggard) [reply eng/ApiListing.targets]: Or do you mean `Export-API.ps1` and if that is the same then how is this different from the CI?  

We also need some docs updated in Contributing.md to tell folks how to do this.

Comment (pakrym) [reply eng/ApiListing.targets]: > how is this different from the CI?

It runs a custom target `GenerateAPIListing` that is not executed by default.

File: eng/service.proj [hunk 8]
```diff
@@ -24,4 +24,14 @@
     <Error Condition="'$(Scope)' == '' and '@(ProjectReference)' == ''"
         Text="No Projects found with patttern [..\sdk\$(ServiceDirectory)\**\*.csproj], please make sure you have passed in the correct ServiceDirectory." />
   </Target>
+
+  <Target Name="GenerateAPIListing"
```
Comment (weshaggard) [inline eng/service.proj]: In order to produce the api you are going to need to do a build so why introduce a new target? Instead I suggest just a property you set.

File: eng/Directory.Build.Data.props [hunk 9]
```diff
@@ -38,9 +38,12 @@
     <IsTestProject Condition="$(MSBuildProjectName.EndsWith('.Tests'))">true</IsTestProject>
     <IsSamplesProject Condition="$(MSBuildProjectName.EndsWith('.Samples'))">true</IsSamplesProject>
     <IsTestSupportProject Condition="'$(IsTestProject)' != 'true' and ($(MSBuildProjectDirectory.Contains('/tests/')) or $(MSBuildProjectDirectory.Contains('\tests\')))">true</IsTestSupportProject>
-    <EnableClientSdkAnalyzers Condition="'$(IsClientLibrary)' == 'true' and '$(IsTestProject)' != 'true' and '$(IsTestSupportProject)' != 'true' and '$(IsSamplesProject)' != 'true'">true</EnableClientSdkAnalyzers>
-    <EnableFxCopAnalyzers Condition="'$(IsClientLibrary)' == 'true' and '$(IsTestProject)' != 'true' and '$(IsTestSupportProject)' != 'true' and '$(IsSamplesProject)' != 'true'">true</EnableFxCopAnalyzers>
+    <IsMainClientLibrary Condition="'$(IsClientLibrary)' == 'true' and '$(IsTestProject)' != 'true' and '$(IsTestSupportProject)' != 'true' and '$(IsSamplesProject)' != 'true'">true</IsMainClientLibrary>
```
Comment (weshaggard) [inline eng/Directory.Build.Data.props]: nit: it would probably be better to call this `IsShippingClientLibrary`.

File: eng/ApiListing.targets (hunk 2 above)
Comment (weshaggard) [reply eng/ApiListing.targets]: Where is it failing the build if there is a diff?

File: eng/Versioning.targets [hunk 10]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <_BuildNumber>$(OfficialBuildId)</_BuildNumber>
     <_BuildNumber Condition="'$(OfficialBuildId)' == ''">$([System.DateTime]::Now.ToString(yyyyMMdd)).1</_BuildNumber>
+    <OriginalVersion>$(Version)</OriginalVersion>
```
Comment (weshaggard) [inline eng/Versioning.targets]: Why is this needed? It seems like you are only using it for release builds and in which case the Version should be unchanged.

File: eng/ApiListing.targets (hunk 2 above)
Comment (pakrym) [reply eng/ApiListing.targets]: https://github.com/Azure/azure-sdk-for-net/pull/8695/files#diff-e393debd7f57cca40fe083227f50bc35

https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/templates/jobs/archetype-sdk-client.yml#L91

File: eng/Versioning.targets (hunk 10 above)
Comment (pakrym) [reply eng/Versioning.targets]: Let me check.

File: eng/ApiListing.targets [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      <_AllTargetFrameworks Remove="@(_AllTargetFrameworks)" />
+      <_AllTargetFrameworks Include="$(TargetFrameworks);$(TargetFramework)" />
+      <_DeduplicatedTargetFramework Include="%(_AllTargetFrameworks.Identity)" />
+    </ItemGroup>
+    
+    <RemoveDir Directories="$(_RefSourceOutputPath)" />
+
+    <Error Condition="'$(GenerateAPIListing)' != 'true' AND !$(OriginalVersion.Contains('preview'))" Text="GenerateAPIListing must be set to true for non release version" />
+
+    <MSBuild Projects="$(MSBuildProjectFullPath)"
```
Comment (weshaggard) [inline eng/ApiListing.targets]: If you do this with a property instead of a custom target you don't need to multiplex over the target frameworks yourself so this target will be much simplier.

Comment (pakrym) [reply eng/ApiListing.targets]: Yeah, but I do other things not just multiplexing.

File: eng/CodeChecks.ps1 [hunk 12]
```diff
@@ -71,7 +76,7 @@ try {
     if ($LastExitCode -ne 0) {
```
Comment (weshaggard) [inline eng/CodeChecks.ps1]: I see this is where you are failing if there are api changes. I'm OK with this approach.

Comment (weshaggard) [inline eng/CodeChecks.ps1]: Why do we need `--ignore-space-at-eol` the generation tools should not be generating different spacing around eol.

File: eng/Versioning.targets (hunk 10 above)
Comment (pakrym) [reply eng/Versioning.targets]: `SkipDevBuildNumber` is not base on configuration, it's set by some CI builds only.

File: eng/ApiListing.targets [hunk 13]
```diff
@@ -0,0 +1,57 @@
+<Project>
```
Comment (weshaggard) [inline eng/ApiListing.targets]: Is there a reason the default targets coming from the package will not work? https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.GenAPI/build/Microsoft.DotNet.GenAPI.targets

File: eng/CodeChecks.ps1 (hunk 12 above)
Comment (pakrym) [reply eng/CodeChecks.ps1]: Is to ignore different line endings that sometimes get generated.

File: eng/ApiListing.targets (hunk 13 above)
Comment (pakrym) [reply eng/ApiListing.targets]: Yes, I wanted more checks that are specific to Azure SDKs and I had experience of being broken by changes in those targets.

File: eng/ApiListing.targets (hunk 1 above)
Comment (AlexGhiondea) [reply eng/ApiListing.targets]: Previous version should be the latest stable version. This should be fine for now.

File: sdk/cognitiveservices/InkRecognizer/api/Azure.AI.InkRecognizer.netstandard2.0.cs [hunk 14]
```diff
@@ -0,0 +1,211 @@
+namespace Azure.AI.InkRecognizer
```
Comment (AlexGhiondea) [inline sdk/cognitiveservices/InkRecognizer/api/Azure.AI.InkRecognizer.netstandard2.0.cs]: I don't think this should be here.

Comment (pakrym) [reply sdk/cognitiveservices/InkRecognizer/api/Azure.AI.InkRecognizer.netstandard2.0.cs]: Do you want me to disable the check for this project? It uses Azure.Core and has Track2 naming,

Comment (AlexGhiondea) [reply sdk/cognitiveservices/InkRecognizer/api/Azure.AI.InkRecognizer.netstandard2.0.cs]: I don't remember shipping a GA for it -- if we have, then we should keep the check.

File: eng/ApiListing.targets (hunk 1 above)
Comment (pakrym) [reply eng/ApiListing.targets]: Yes, but there are two goals:
1. Being aware of what API changes are happening - they are often hard to discern from large code change
2. Making sure they are not breaking

File: sdk/cognitiveservices/InkRecognizer/api/Azure.AI.InkRecognizer.netstandard2.0.cs (hunk 14 above)
Comment (weshaggard) [reply sdk/cognitiveservices/InkRecognizer/api/Azure.AI.InkRecognizer.netstandard2.0.cs]: We plan to actually delete this project from the repo. So do whatever is easiest for this.

Comment (pakrym) [reply sdk/cognitiveservices/InkRecognizer/api/Azure.AI.InkRecognizer.netstandard2.0.cs]: We haven't shipped some other assemblies I have this enabled on. It's nice to be able to see what changes are being done to public API as part of PR.

File: eng/ApiListing.targets (hunk 11 above)
Comment (weshaggard) [reply eng/ApiListing.targets]: What other things are you doing? I'd prefer us not try to mimic this multiplexing as it is complicated to keep working.

File: eng/ApiListing.targets (hunk 13 above)
Comment (weshaggard) [reply eng/ApiListing.targets]: There should be enough properties to override the specific things we want for us and if we can do that with a couple property changes we should do it as opposed to duplicating it.

**Comment by AlexGhiondea** [comment]
This is great!!

Any reason to not enable this by default in our builds? How much does it add to the build time?

I would like the system to be as automated as possible as this needs to 'just work', in particular for our partner teams. Right now this makes it easier for our team to catch changes, but still requires people to catch it in code reviews.

Going forward, I think these are the things we want to enforce/allow:
- API changes to stable versions of the API are NOT allowed
  -  If the API is present in a stable package, it can't have breaking changes (I believe there are some changes that we can allow).
- API changes to preview versions of the API are allowed.
  -  This means that we can break APIs (rename, remove, etc) as long as we haven't shipped that library in a stable package.

File: eng/Versioning.targets (hunk 10 above)
Comment (weshaggard) [reply eng/Versioning.targets]: Correct but if we are setting a dev version we don't have to error, in fact for our checkcode script should never be passing that.

**Comment by pakrym** [comment]
> Any reason to not enable this by default in our builds? How much does it add to the build time?

It is enabled by default in our builds in the Analyze step.

>I would like the system to be as automated as possible as this needs to 'just work', in particular for our partner teams. Right now this makes it easier for our team to catch changes, but still requires people to catch it in code reviews.

The intent was to be able to at least clearly see API changes. We'll add more rules and automation later.

File: eng/ApiListing.targets (hunk 11 above)
Comment (pakrym) [reply eng/ApiListing.targets]: For example removing the `api`  directory if GenerateAPIListing is disabled to make sure the setting and the listing is in sync.

File: eng/Versioning.targets (hunk 10 above)
Comment (pakrym) [reply eng/Versioning.targets]: I wan't a check to verify that GenerateAPIListing isn't set to false in projects with release version set. How do I do it without storing the original version?

File: eng/ApiListing.targets (hunk 11 above)
Comment (weshaggard) [reply eng/ApiListing.targets]: That can be done as another target. You can easily add that as a before target for the shared target.

File: eng/Versioning.targets (hunk 10 above)
Comment (weshaggard) [reply eng/Versioning.targets]: You can use just the `Version` property. If it is not a dev build it will be exactly what you want and if it is a dev build then we aren't releasing it so it doesn't matter.

Comment (pakrym) [reply eng/Versioning.targets]: > You can use just the Version property. If it is not a dev build it will be exactly what you want and if it is a dev build then we aren't releasing it so it doesn't matter.

I want it to fail in Analyze target in PR and Export-API.ps1 call . We are not releasing at that point.

File: eng/ApiListing.targets (hunk 11 above)
Comment (pakrym) [reply eng/ApiListing.targets]: Done

File: eng/ApiListing.targets (hunk 13 above)
Comment (pakrym) [reply eng/ApiListing.targets]: Done

File: eng/service.proj (hunk 8 above)
Comment (pakrym) [reply eng/service.proj]: Done

File: eng/Versioning.targets (hunk 10 above)
Comment (weshaggard) [reply eng/Versioning.targets]: We are also not passing the dev build option during this CodeChecks.ps1 you are in control of that so it shouldn't ever be in that state. In fact if you want Export-APIs.ps1 to never be in that case you can explicitly pass `/p:SkipDevBuildNumber=false`

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -115,6 +115,10 @@ If for any reason there is an update to the build tools, you will then need to f
 3. Invoke `dotnet test eng\service.proj --filter TestCategory!=Live`
    <br/><br/>
 
+## Public API changes
+
+If you make a public API change `eng\Export-API.ps1` script has to be run to update public API listings.
```
Comment (weshaggard) [inline CONTRIBUTING.md]: "run `.\eng\Export-API.ps1". 

Any reason to not just call be with the property?

Comment (pakrym) [reply CONTRIBUTING.md]: I think it's easier for people to just invoke one repo level script.

File: eng/CodeChecks.ps1 (hunk 12 above)
Comment (weshaggard) [reply eng/CodeChecks.ps1]: We should update our generation tools to make that not an issue but I'm OK with the diff ignoring it for now.

**Review by weshaggard — APPROVED** [review]
Thanks for taking this task on and implementing.

**Comment by pakrym** [comment]
Flaky test: https://github.com/Azure/azure-sdk-for-net/issues/8732
